File: data_baseline/openrouter__qwen__qwen3.7-plus/haskell/rep6/Main.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Control.Concurrent.STM (TVar, atomically, readTVar, writeTVar, readTVarIO, newTVarIO)
import Control.Monad.IO.Class (liftIO)
import Data.Char (isAsciiLower, isAsciiUpper, isDigit)
import Data.List (find, sortOn)
import Data.Maybe (fromMaybe)
import Data.Text (Text)
import qualified Data.Text as T
import Data.Text.Encoding (decodeUtf8)
import qualified Data.Text.Lazy as TL
import Data.Time (UTCTime, getCurrentTime, formatTime, defaultTimeLocale)
import Data.UUID (toString)
import Data.UUID.V4 (nextRandom)
import GHC.Generics (Generic)
import Web.Scotty (scotty, post, get, put, delete, jsonData, request, addHeader, status, json, ActionM)
import Network.HTTP.Types (status400, status401, status404, status409, status200, status201, status204)
import qualified Network.Wai as Wai
import qualified Data.ByteString.Char8 as B8
import qualified Data.Map.Strict as Map
import Data.Aeson
import System.Environment (getArgs)
import Text.Read (readMaybe)

data User = User
  { uId :: Int
  , uUsername :: Text
  , uPassword :: Text
  } deriving (Show, Generic)

instance ToJSON User where
  toJSON (User i u _) = object ["id" .= i, "username" .= u]

data Todo = Todo
  { tId :: Int
  , tUserId :: Int
  , tTitle :: Text
  , tDescription :: Text
  , tCompleted :: Bool
  , tCreatedAt :: UTCTime
  , tUpdatedAt :: UTCTime
  } deriving (Show, Generic)

instance ToJSON Todo where
  toJSON (Todo i _ title desc comp cTime uTime) = object
    [ "id" .= i
    , "title" .= title
    , "description" .= desc
    , "completed" .= comp
    , "created_at" .= formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%SZ" cTime
    , "updated_at" .= formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%SZ" uTime
    ]

data AppState = AppState
  { nextUserId :: Int
  , nextTodoId :: Int
  , users :: Map.Map Int User
  , userByName :: Map.Map Text Int
  , todos :: Map.Map Int Todo
  , sessions :: Map.Map Text Int
  } deriving (Show)

initialState :: AppState
initialState = AppState
  { nextUserId = 1
  , nextTodoId = 1
  , users = Map.empty
  , userByName = Map.empty
  , todos = Map.empty
  , sessions = Map.empty
  }

isValidUsername :: Text -> Bool
isValidUsername t =
  let len = T.length t
  in len >= 3 && len <= 50 && T.all (\c -> isAsciiUpper c || isAsciiLower c || isDigit c || c == '_') t

isValidPassword :: Text -> Bool
isValidPassword t = T.length t >= 8

data RegisterReq = RegisterReq
  { regUsername :: Text
  , regPassword :: Text
  } deriving (Generic)

instance FromJSON RegisterReq where
  parseJSON = withObject "RegisterReq" $ \v -> RegisterReq
    <$> v .: "username"
    <*> v .: "password"

data LoginReq = LoginReq
  { logUsername :: Text
  , logPassword :: Text
  } deriving (Generic)

instance FromJSON LoginReq where
  parseJSON = withObject "LoginReq" $ \v -> LoginReq
    <$> v .: "username"
    <*> v .: "password"

data PasswordReq = PasswordReq
  { reqOldPassword :: Text
  , reqNewPassword :: Text
  } deriving (Generic)

instance FromJSON PasswordReq where
  parseJSON = withObject "PasswordReq" $ \v -> PasswordReq
    <$> v .: "old_password"
    <*> v .: "new_password"

data TodoReq = TodoReq
  { reqTitle :: Text
  , reqDescription :: Maybe Text
  } deriving (Generic)

instance FromJSON TodoReq where
  parseJSON = withObject "TodoReq" $ \v -> TodoReq
    <$> v .: "title"
    <*> v .:? "description"

data UpdateTodoReq = UpdateTodoReq
  { upTitle :: Maybe Text
  , upDescription :: Maybe Text
  , upCompleted :: Maybe Bool
  } deriving (Generic)

instance FromJSON UpdateTodoReq where
  parseJSON = withObject "UpdateTodoReq" $ \v -> UpdateTodoReq
    <$> v .:? "title"
    <*> v .:? "description"
    <*> v .:? "completed"

getSid :: ActionM (Maybe Text)
getSid = do
  req <- request
  let headers = Wai.requestHeaders req
  let mCookie = lookup "cookie" headers
  case mCookie of
    Nothing -> return Nothing
    Just cookieBS -> 
      let cookies = B8.split ';' cookieBS
          match = find (\c -> "session_id=" `B8.isPrefixOf` B8.dropWhile (== ' ') c) cookies
      in case match of
           Nothing -> return Nothing
           Just m -> return (Just (T.pack (B8.unpack (B8.drop 11 (B8.dropWhile (== ' ') m)))))

getPathParam :: ActionM Text
getPathParam = do
  req <- request
  let path = decodeUtf8 (Wai.rawPathInfo req)
  let parts = T.split (== '/') path
  case reverse (filter (not . T.null) parts) of
    (p:_) -> return p
    [] -> return ""

main :: IO ()
main = do
  args <- getArgs
  let portStr = case args of
        ["--port", p] -> p
        _ -> "3000"
  let port = read portStr :: Int
  appStateVar <- newTVarIO initialState
  
  let auth' :: (Int -> Text -> ActionM ()) -> ActionM ()
      auth' action = do
        mSid <- getSid
        case mSid of
          Nothing -> do
            status status401
            json (object ["error" .= ("Authentication required" :: Text)])
          Just sid -> do
            st <- liftIO $ readTVarIO appStateVar
            case Map.lookup sid (sessions st) of
              Nothing -> do
                status status401
                json (object ["error" .= ("Authentication required" :: Text)])
              Just uid -> action uid sid

  scotty port $ do
    post "/register" $ do
      req <- jsonData :: ActionM RegisterReq
      let uname = regUsername req
      let pword = regPassword req
      if not (isValidUsername uname)
        then do
          status status400
          json (object ["error" .= ("Invalid username" :: Text)])
        else if not (isValidPassword pword)
          then do
            status status400
            json (object ["error" .= ("Password too short" :: Text)])
          else do
            res <- liftIO $ atomically $ do
              st <- readTVar appStateVar
              if Map.member uname (userByName st)
                then return (Left ("Username already exists" :: Text))
                else do
                  let uid = nextUserId st
                  let u = User uid uname pword
                  let st' = st
                        { nextUserId = uid + 1
                        , users = Map.insert uid u (users st)
                        , userByName = Map.insert uname uid (userByName st)
                        }
                  writeTVar appStateVar st'
                  return (Right (uid, uname))
            case res of
              Left err -> do
                status status409
                json (object ["error" .= err])
              Right (uid, uname') -> do
                status status201
                json (object ["id" .= uid, "username" .= uname'])

    post "/login" $ do
      req <- jsonData :: ActionM LoginReq
      let uname = logUsername req
      let pword = logPassword req
      res <- liftIO $ atomically $ do
        st <- readTVar appStateVar
        case Map.lookup uname (userByName st) of
          Nothing -> return (Nothing :: Maybe Int)
          Just uid -> case Map.lookup uid (users st) of
            Nothing -> return Nothing
            Just u -> if uPassword u == pword
              then return (Just uid)
              else return Nothing
      case res of
        Nothing -> do
          status status401
          json (object ["error" .= ("Invalid credentials" :: Text)])
        Just uid -> do
          token <- liftIO $ toString <$> nextRandom
          liftIO $ atomically $ do
            st <- readTVar appStateVar
            let st' = st { sessions = Map.insert (T.pack token) uid (sessions st) }
            writeTVar appStateVar st'
          addHeader "Set-Cookie" (TL.pack ("session_id=" ++ token ++ "; Path=/; HttpOnly"))
          status status200
          json (object ["id" .= (uid :: Int), "username" .= uname])

    post "/logout" $ do
      auth' $ \_uid sid -> do
        liftIO $ atomically $ do
          st <- readTVar appStateVar
          let st' = st { sessions = Map.delete sid (sessions st) }
          writeTVar appStateVar st'
        status status200
        json (object [] :: Value)

    get "/me" $ do
      auth' $ \uid _sid -> do
        st <- liftIO $ readTVarIO appStateVar
        case Map.lookup uid (users st) of
          Nothing -> do
            status status401
            json (object ["error" .= ("Authentication required" :: Text)])
          Just u -> do
            status status200
            json (object ["id" .= uId u, "username" .= uUsername u])

    put "/password" $ do
      auth' $ \uid _sid -> do
        req <- jsonData :: ActionM PasswordReq
        let oldP = reqOldPassword req
        let newP = reqNewPassword req
        if not (isValidPassword newP)
          then do
            status status400
            json (object ["error" .= ("Password too short" :: Text)])
          else do
            st <- liftIO $ readTVarIO appStateVar
            case Map.lookup uid (users st) of
              Nothing -> do
                status status401
                json (object ["error" .= ("Invalid credentials" :: Text)])
              Just u -> if uPassword u /= oldP
                then do
                  status status401
                  json (object ["error" .= ("Invalid credentials" :: Text)])
                else do
                  liftIO $ atomically $ do
                    st' <- readTVar appStateVar
                    let u' = u { uPassword = newP }
                    let st'' = st' { users = Map.insert uid u' (users st') }
                    writeTVar appStateVar st''
                  status status200
                  json (object [] :: Value)

    get "/todos" $ do
      auth' $ \uid _sid -> do
        st <- liftIO $ readTVarIO appStateVar
        let userTodos = Map.elems $ Map.filter (\t -> tUserId t == uid) (todos st)
        let sortedTodos = sortOn tId userTodos
        status status200
        json sortedTodos

    post "/todos" $ do
      auth' $ \uid _sid -> do
        req <- jsonData :: ActionM TodoReq
        let t = reqTitle req
        if T.null t
          then do
            status status400
            json (object ["error" .= ("Title is required" :: Text)])
          else do
            let desc = fromMaybe "" (reqDescription req)
            now <- liftIO getCurrentTime
            res <- liftIO $ atomically $ do
              st' <- readTVar appStateVar
              let tid = nextTodoId st'
              let newTodo = Todo tid uid t desc False now now
              let st'' = st'
                    { nextTodoId = tid + 1
                    , todos = Map.insert tid newTodo (todos st')
                    }
              writeTVar appStateVar st''
              return newTodo
            status status201
            json res

    get "/todos/:id" $ do
      auth' $ \uid _sid -> do
        tidTxt <- getPathParam
        case readMaybe (T.unpack tidTxt) of
          Nothing -> do
            status status404
            json (object ["error" .= ("Todo not found" :: Text)])
          Just tid -> do
            st <- liftIO $ readTVarIO appStateVar
            case Map.lookup tid (todos st) of
              Just t | tUserId t == uid -> do
                status status200
                json t
              _ -> do
                status status404
                json (object ["error" .= ("Todo not found" :: Text)])

    put "/todos/:id" $ do
      auth' $ \uid _sid -> do
        tidTxt <- getPathParam
        case readMaybe (T.unpack tidTxt) of
          Nothing -> do
            status status404
            json (object ["error" .= ("Todo not found" :: Text)])
          Just tid -> do
            req <- jsonData :: ActionM UpdateTodoReq
            case upTitle req of
              Just "" -> do
                status status400
                json (object ["error" .= ("Title is required" :: Text)])
              _ -> do
                now <- liftIO getCurrentTime
                res <- liftIO $ atomically $ do
                  st' <- readTVar appStateVar
                  case Map.lookup tid (todos st') of
                    Just t | tUserId t == uid -> do
                      let newTitle = fromMaybe (tTitle t) (upTitle req)
                      let newDesc = fromMaybe (tDescription t) (upDescription req)
                      let newComp = fromMaybe (tCompleted t) (upCompleted req)
                      let newTodo = t { tTitle = newTitle, tDescription = newDesc, tCompleted = newComp, tUpdatedAt = now }
                      let st'' = st' { todos = Map.insert tid newTodo (todos st') }
                      writeTVar appStateVar st''
                      return (Right newTodo :: Either Text Todo)
                    _ -> return (Left "Todo not found")
                case res of
                  Left _ -> do
                    status status404
                    json (object ["error" .= ("Todo not found" :: Text)])
                  Right t -> do
                    status status200
                    json t

    delete "/todos/:id" $ do
      auth' $ \uid _sid -> do
        tidTxt <- getPathParam
        case readMaybe (T.unpack tidTxt) of
          Nothing -> do
            status status404
            json (object ["error" .= ("Todo not found" :: Text)])
          Just tid -> do
            res <- liftIO $ atomically $ do
              st' <- readTVar appStateVar
              case Map.lookup tid (todos st') of
                Just t | tUserId t == uid -> do
                  let st'' = st' { todos = Map.delete tid (todos st') }
                  writeTVar appStateVar st''
                  return (True :: Bool)
                _ -> return False
            if res
              then status status204
              else do
                status status404
                json (object ["error" .= ("Todo not found" :: Text)])

File: data_baseline/openrouter__qwen__qwen3.7-plus/rust/rep6/src/main.rs
use std::sync::{Arc, Mutex};
use std::collections::HashMap;
use axum::{
    extract::{Path, State},
    http::{header::SET_COOKIE, HeaderMap, StatusCode},
    response::{IntoResponse, Json},
    routing::{delete, get, post, put},
    Router,
};
use serde::{Deserialize, Serialize};
use serde_json::json;
use uuid::Uuid;
use chrono::Utc;
use regex::Regex;
use bcrypt::{hash, verify, DEFAULT_COST};

#[derive(Clone)]
struct AppState {
    users: Arc<Mutex<Vec<User>>>,
    todos: Arc<Mutex<Vec<Todo>>>,
    sessions: Arc<Mutex<HashMap<String, i32>>>,
    next_user_id: Arc<Mutex<i32>>,
    next_todo_id: Arc<Mutex<i32>>,
}

#[derive(Clone)]
struct User {
    id: i32,
    username: String,
    password_hash: String,
}

#[derive(Serialize, Deserialize, Clone)]
struct Todo {
    id: i32,
    title: String,
    description: String,
    completed: bool,
    created_at: String,
    updated_at: String,
    #[serde(skip)]
    user_id: i32,
}

struct UserInfo {
    id: i32,
    username: String,
}

fn hash_password(password: &str) -> String {
    hash(password, DEFAULT_COST).unwrap()
}

fn verify_password(password: &str, hash_str: &str) -> bool {
    verify(password, hash_str).unwrap_or(false)
}

fn get_user_from_headers(
    headers: &HeaderMap,
    state: &AppState,
) -> Result<UserInfo, (StatusCode, Json<serde_json::Value>)> {
    let cookie_header = headers.get("cookie");
    if let Some(cookie) = cookie_header {
        let cookie_str = cookie.to_str().unwrap_or("");
        for c in cookie_str.split(';') {
            let c = c.trim();
            if let Some(session_id) = c.strip_prefix("session_id=") {
                let sessions = state.sessions.lock().unwrap();
                if let Some(&user_id) = sessions.get(session_id) {
                    let users = state.users.lock().unwrap();
                    if let Some(user) = users.iter().find(|u| u.id == user_id) {
                        return Ok(UserInfo {
                            id: user.id,
                            username: user.username.clone(),
                        });
                    }
                }
            }
        }
    }
    Err((
        StatusCode::UNAUTHORIZED,
        Json(json!({"error": "Authentication required"})),
    ))
}

#[derive(Deserialize)]
struct RegisterRequest {
    username: Option<String>,
    password: Option<String>,
}

async fn register(
    State(state): State<AppState>,
    Json(req): Json<RegisterRequest>,
) -> Result<(StatusCode, Json<serde_json::Value>), (StatusCode, Json<serde_json::Value>)> {
    let username = req.username.unwrap_or_default();
    let password = req.password.unwrap_or_default();

    let re = Regex::new(r"^[a-zA-Z0-9_]+$").unwrap();
    if username.len() < 3 || username.len() > 50 || !re.is_match(&username) {
        return Err((StatusCode::BAD_REQUEST, Json(json!({"error": "Invalid username"}))));
    }

    if password.len() < 8 {
        return Err((StatusCode::BAD_REQUEST, Json(json!({"error": "Password too short"}))));
    }

    let mut users = state.users.lock().unwrap();
    if users.iter().any(|u| u.username == username) {
        return Err((StatusCode::CONFLICT, Json(json!({"error": "Username already exists"}))));
    }

    let id = *state.next_user_id.lock().unwrap();
    *state.next_user_id.lock().unwrap() += 1;

    let password_hash = hash_password(&password);
    users.push(User {
        id,
        username: username.clone(),
        password_hash,
    });

    Ok((StatusCode::CREATED, Json(json!({
        "id": id,
        "username": username
    }))))
}

#[derive(Deserialize)]
struct LoginRequest {
    username: Option<String>,
    password: Option<String>,
}

async fn login(
    State(state): State<AppState>,
    Json(req): Json<LoginRequest>,
) -> Result<(HeaderMap, Json<serde_json::Value>), (StatusCode, Json<serde_json::Value>)> {
    let username = req.username.unwrap_or_default();
    let password = req.password.unwrap_or_default();

    let users = state.users.lock().unwrap();
    let user_opt = users.iter().find(|u| u.username == username).cloned();
    drop(users);

    if let Some(user) = user_opt {
        if verify_password(&password, &user.password_hash) {
            let session_id = Uuid::new_v4().to_string();
            state.sessions.lock().unwrap().insert(session_id.clone(), user.id);

            let mut headers = HeaderMap::new();
            headers.insert(SET_COOKIE, format!("session_id={}; Path=/; HttpOnly", session_id).parse().unwrap());
            
            return Ok((headers, Json(json!({
                "id": user.id,
                "username": user.username
            }))));
        }
    }

    Err((StatusCode::UNAUTHORIZED, Json(json!({"error": "Invalid credentials"}))))
}

async fn logout(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> Result<Json<serde_json::Value>, (StatusCode, Json<serde_json::Value>)> {
    let _user = get_user_from_headers(&headers, &state)?;
    
    let cookie_header = headers.get("cookie");
    if let Some(cookie) = cookie_header {
        let cookie_str = cookie.to_str().unwrap_or("");
        for c in cookie_str.split(';') {
            let c = c.trim();
            if let Some(session_id) = c.strip_prefix("session_id=") {
                state.sessions.lock().unwrap().remove(session_id);
                break;
            }
        }
    }

    Ok(Json(json!({})))
}

async fn get_me(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> Result<Json<serde_json::Value>, (StatusCode, Json<serde_json::Value>)> {
    let user = get_user_from_headers(&headers, &state)?;
    Ok(Json(json!({
        "id": user.id,
        "username": user.username
    })))
}

#[derive(Deserialize)]
struct PasswordUpdateRequest {
    old_password: Option<String>,
    new_password: Option<String>,
}

async fn update_password(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(req): Json<PasswordUpdateRequest>,
) -> Result<Json<serde_json::Value>, (StatusCode, Json<serde_json::Value>)> {
    let user = get_user_from_headers(&headers, &state)?;
    
    let old_password = req.old_password.unwrap_or_default();
    let new_password = req.new_password.unwrap_or_default();

    if new_password.len() < 8 {
        return Err((StatusCode::BAD_REQUEST, Json(json!({"error": "Password too short"}))));
    }

    let mut users = state.users.lock().unwrap();
    let user_idx = users.iter().position(|u| u.id == user.id).unwrap();
    
    if !verify_password(&old_password, &users[user_idx].password_hash) {
        return Err((StatusCode::UNAUTHORIZED, Json(json!({"error": "Invalid credentials"}))));
    }

    users[user_idx].password_hash = hash_password(&new_password);

    Ok(Json(json!({})))
}

async fn get_todos(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> Result<Json<serde_json::Value>, (StatusCode, Json<serde_json::Value>)> {
    let user = get_user_from_headers(&headers, &state)?;
    
    let mut todos: Vec<_> = state.todos.lock().unwrap()
        .iter()
        .filter(|t| t.user_id == user.id)
        .cloned()
        .collect();
    
    todos.sort_by_key(|t| t.id);

    Ok(Json(serde_json::to_value(todos).unwrap()))
}

#[derive(Deserialize)]
struct CreateTodoRequest {
    title: Option<String>,
    description: Option<String>,
}

async fn create_todo(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(req): Json<CreateTodoRequest>,
) -> Result<(StatusCode, Json<serde_json::Value>), (StatusCode, Json<serde_json::Value>)> {
    let user = get_user_from_headers(&headers, &state)?;
    
    let title = req.title.unwrap_or_default();
    if title.is_empty() {
        return Err((StatusCode::BAD_REQUEST, Json(json!({"error": "Title is required"}))));
    }

    let description = req.description.unwrap_or_default();
    let now = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();

    let id = *state.next_todo_id.lock().unwrap();
    *state.next_todo_id.lock().unwrap() += 1;

    let todo = Todo {
        id,
        user_id: user.id,
        title,
        description,
        completed: false,
        created_at: now.clone(),
        updated_at: now,
    };

    state.todos.lock().unwrap().push(todo.clone());

    Ok((StatusCode::CREATED, Json(serde_json::to_value(todo).unwrap())))
}

async fn get_todo(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(todo_id): Path<i32>,
) -> Result<Json<serde_json::Value>, (StatusCode, Json<serde_json::Value>)> {
    let user = get_user_from_headers(&headers, &state)?;
    
    let todos = state.todos.lock().unwrap();
    let todo = todos.iter().find(|t| t.id == todo_id && t.user_id == user.id).cloned();
    
    match todo {
        Some(t) => Ok(Json(serde_json::to_value(t).unwrap())),
        None => Err((StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"})))),
    }
}

#[derive(Deserialize)]
struct UpdateTodoRequest {
    title: Option<String>,
    description: Option<String>,
    completed: Option<bool>,
}

async fn update_todo(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(todo_id): Path<i32>,
    Json(req): Json<UpdateTodoRequest>,
) -> Result<Json<serde_json::Value>, (StatusCode, Json<serde_json::Value>)> {
    let user = get_user_from_headers(&headers, &state)?;
    
    if let Some(ref title) = req.title {
        if title.is_empty() {
            return Err((StatusCode::BAD_REQUEST, Json(json!({"error": "Title is required"}))));
        }
    }

    let mut todos = state.todos.lock().unwrap();
    let todo_idx = todos.iter().position(|t| t.id == todo_id && t.user_id == user.id);

    match todo_idx {
        Some(idx) => {
            if let Some(title) = req.title {
                todos[idx].title = title;
            }
            if let Some(description) = req.description {
                todos[idx].description = description;
            }
            if let Some(completed) = req.completed {
                todos[idx].completed = completed;
            }
            todos[idx].updated_at = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();
            
            Ok(Json(serde_json::to_value(todos[idx].clone()).unwrap()))
        }
        None => Err((StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"})))),
    }
}

async fn delete_todo(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(todo_id): Path<i32>,
) -> Result<impl IntoResponse, (StatusCode, Json<serde_json::Value>)> {
    let user = get_user_from_headers(&headers, &state)?;
    
    let mut todos = state.todos.lock().unwrap();
    let todo_idx = todos.iter().position(|t| t.id == todo_id && t.user_id == user.id);

    match todo_idx {
        Some(idx) => {
            todos.remove(idx);
            Ok(StatusCode::NO_CONTENT)
        }
        None => Err((StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"})))),
    }
}

#[tokio::main]
async fn main() {
    let args: Vec<String> = std::env::args().collect();
    let mut port = 3000;
    for i in 0..args.len() {
        if args[i] == "--port" && i + 1 < args.len() {
            port = args[i + 1].parse().unwrap_or(3000);
        }
    }

    let state = AppState {
        users: Arc::new(Mutex::new(Vec::new())),
        todos: Arc::new(Mutex::new(Vec::new())),
        sessions: Arc::new(Mutex::new(HashMap::new())),
        next_user_id: Arc::new(Mutex::new(1)),
        next_todo_id: Arc::new(Mutex::new(1)),
    };

    let app = Router::new()
        .route("/register", post(register))
        .route("/login", post(login))
        .route("/logout", post(logout))
        .route("/me", get(get_me))
        .route("/password", put(update_password))
        .route("/todos", get(get_todos))
        .route("/todos", post(create_todo))
        .route("/todos/:id", get(get_todo))
        .route("/todos/:id", put(update_todo))
        .route("/todos/:id", delete(delete_todo))
        .with_state(state);

    let addr = format!("0.0.0.0:{}", port);
    let listener = tokio::net::TcpListener::bind(&addr).await.unwrap();
    println!("Listening on {}", addr);
    axum::serve(listener, app).await.unwrap();
}
